File: python/select_next_word_ranking.py
import numpy as np
from scipy.stats import multivariate_normal
import select_permutations as select
from math import factorial, log2
import scipy.io as sio
def select_next_word_ranking(previous_words, previous_index, list_embeddings, mu, sigma, noise_factor, scale, 
                             N_particles=None, max_permutations=500):
    N_prev = previous_words.shape[1]    

    if N_particles is None:
        N_particles = max(200, int(2e4 / factorial(N_prev + 1)))

    N_words = list_embeddings.shape[1]

    # Define probability functions
    def p_y1_xandtheta(theta, S):
        return 1 / (1 + np.exp(- (noise_factor * (theta.T @ S))))
    
    def softmax_num(theta, S):
        return np.exp(scale * (theta.T @ S))
    mat_data = sio.loadmat(r'C:\Users\yslee\OneDrive\Desktop\Research\main\test_results_matlab.mat')
    particles = mat_data['particles_mat']
    # Sample and normalize particles
    # particles = multivariate_normal.rvs(mean=mu.flatten(), cov=sigma, size=N_particles).T
    # particles /= np.linalg.norm(particles, axis=0, keepdims=True)

    # Compute p_y1
    p_y1_words = np.zeros((N_particles, N_prev + 1, N_words), dtype=np.float32)
    p_y1_words[:, :N_prev, :] = np.expand_dims(p_y1_xandtheta(particles, previous_words), axis=2)
    p_y1_words[:, N_prev, :] = p_y1_xandtheta(particles, list_embeddings)

    p_y1_all1 = np.prod(p_y1_words, axis=1)
    p_y1_allm1 = np.prod(1 - p_y1_words, axis=1)

    temp_options = N_prev + 2
    p_y = np.zeros((N_particles, temp_options, N_words), dtype=np.float32)
    p_y[:, 0, :] = p_y1_all1
    p_y[:, -1, :] = p_y1_allm1

    print("\n[DEBUG] p_y1_all1 (Shape:", p_y1_all1.shape, "):", p_y1_all1[:5, :5])  # Debug First 5 rows and columns
    print("[DEBUG] p_y1_allm1 (Shape:", p_y1_allm1.shape, "):", p_y1_allm1[:5, :5])

    # Compute softmax values
    softmax_num_previous_words = softmax_num(particles, previous_words)
    softmax_num_words = softmax_num(particles, list_embeddings)

    # Compute heuristic with numerical stability
    p_xy_safe = np.clip(p_y, 1e-12, 1)  # Prevent zero values
    xylogxy_all = np.sum(p_y * np.log2(p_xy_safe), axis=1)
    E_H_particle = -np.mean(xylogxy_all, axis=0)

    E_p_xy = np.mean(p_y, axis=0)  # Fix entropy normalization
    E_p_xy_safe = np.clip(E_p_xy, 1e-12, 1)  # Prevent log(0)
    H_E_particle = -np.sum(E_p_xy * np.log2(E_p_xy_safe), axis=0)

    H_ratio = H_E_particle - E_H_particle

    print("\n[DEBUG] H_ratio (Top 10 Values):", np.sort(H_ratio)[-5:])  # Print the 10 largest entropy values
    print("[DEBUG] Max H_ratio:", np.max(H_ratio))

    # Select word with maximum entropy heuristic
    max_score = float(np.max(H_ratio))
    sorted_indices = np.argsort(-H_ratio)  # Sort in descending order
    new_index = sorted_indices[0]  # Select highest-ranked word directly

    # Check that new_index does not exceed available words
    new_index = min(new_index, list_embeddings.shape[1])  
    new_index = new_index + 1

    print("\n[DEBUG] Sorted Indices (First 10):", sorted_indices[:10])
    print("[DEBUG] Selected new_index BEFORE fix:", new_index)

    print("[DEBUG] Selected new_index AFTER fix:", new_index)

    # Select the corresponding word
    word = list_embeddings[:, new_index - 1][:, np.newaxis]  # Correct word selection
    list_embeddings = np.delete(list_embeddings, new_index - 1, axis=1)  # Ensure correct removal

    # Update index
    index = np.concatenate((previous_index, [new_index]))  # Ensure MATLAB-like indexing

    words = np.concatenate((previous_words, word), axis=1)

    print("\n[DEBUG] Final Index:", index)
    print("[DEBUG] Final Max Score:", max_score)
    
    return words, index, list_embeddings, max_score

File: matlab/select_next_word_ranking.m
function [words, index, list_embeddings, max_score] = select_next_word_ranking(previous_words, ...
    previous_index, list_embeddings, mu, sigma, noise_factor, scale, N_particles, max_permuations)
%SELECT NEXT WORD RANKING Select greedily the next word to add to the set for query based on
%heuristic of entropy ratios: H(Expectation(Y)) / E(H(Y)) for query of word ranking.
    N_prev = size(previous_words, 2);
    if nargin < 8
        N_particles = max(200, int32(2e4 / factorial(N_prev + 1)));
    end
    if nargin < 9
        max_permuations = 500;
    end
    
    N_words = size(list_embeddings, 2);
    p_y1_xandtheta = @(theta, S) 1 ./ (1 + exp(- noise_factor * (theta' * S))); 
    softmax_num = @(theta, S) exp((scale * (theta' * S)));
    % Sample and normalize particles
    particles = single(mvnrnd(mu, sigma, N_particles)');
    particles = particles ./ vecnorm(particles);

    %% Py
    p_y1_words = zeros(N_particles, N_prev + 1, N_words, 'single');
    p_y1_previous_words = p_y1_xandtheta(particles, previous_words);
    p_y1_particles_words = p_y1_xandtheta(particles, list_embeddings);
    p_y1_previous_words_rep = permute(repmat(p_y1_previous_words, [1, 1, N_words]), [1, 2, 3]);
    p_y1_words(:, 1:N_prev, :) = p_y1_previous_words_rep;
    p_y1_words(:, N_prev + 1 , :) = p_y1_particles_words;
    p_y1_all1 = squeeze(prod(p_y1_words, 2));
    p_y1_allm1 = squeeze(prod(1 - p_y1_words, 2));
    clearvars p_y1_previous_words p_y1_previous_words_rep p_y1_particles_words
    
    temp_options = N_prev + 2;
    p_y = zeros(N_particles, temp_options, N_words, 'single');
    %{
    if max(max(p_y1_all1)) < 0.05 && mean(mean(p_y1_all1)) < 0.010
        p_y(:, 1, :) = [];
        temp_options = temp_options - 1;
    else
        p_y(:, 1, :) = p_y1_all1;
    end
    if max(max(p_y1_allm1)) < 0.05 && mean(mean(p_y1_allm1)) < 0.010
        p_y(:, end, :) = [];
        temp_options = temp_options - 1;
    else
        p_y(:, end, :) = p_y1_allm1;
    end
    %}
    p_y(:, 1, :) = p_y1_all1;
    p_y(:, end, :) = p_y1_allm1;
    clearvars p_y1_all1 p_y1_allm1 

    softmax_num_previous_words = softmax_num(particles, previous_words);
    softmax_num_words = softmax_num(particles, list_embeddings);
    [~, top_option] = sort([mean(softmax_num_previous_words, 1), mean(mean(softmax_num_words))], 2, 'descend');
    ranking_options = select_permutations(N_prev, top_option, max_permuations);% 1e3);
    N_rankings = size(ranking_options, 1);
    %{
    if N_prev >=1
        disp(['N_prev = ' , num2str(N_prev) ,', N_particles = ' ,...
            num2str(N_particles) , ' and N_rankings ' ,num2str(N_rankings)])
    end
    %}
    p_y = repmat(p_y, [1, N_rankings, 1]);  
    clearvars top_options permutations

    for last_pos = 1:N_prev
        pos_ind = ranking_options(:, 1:last_pos);
        neg_ind = ranking_options(:, last_pos+1:end);
        for iter = 1:N_rankings
            p_y1_last_pos = prod(p_y1_words(:, pos_ind(iter, :), :), 2) .*...
                            prod(1 - p_y1_words(:, neg_ind(iter, :), :), 2); % Size N_particles, 1, N_words
            p_y(:, (iter-1)*temp_options + last_pos + 1, :) = p_y1_last_pos; % Size N_particles, N_rankings, N_words
        end
    end
    clearvars p_y1_last_pos p_y1_words
    
    %% P_x
    p_x_particles_first = zeros(N_particles, N_prev + 1, N_words, 'single');
    p_x_particles_first(:, 1:N_prev, :) = repmat(softmax_num_previous_words, [1, 1, N_words]);
    p_x_particles_first(:, end, :) = softmax_num_words;
    p_x_rankings = zeros(N_particles, N_rankings, N_words, 'single');
    clearvars softmax_num_previous_words softmax_num_words 

    for rr = 1:N_rankings
        current_ranking = ranking_options(rr, :);
        p_rr = ones(N_particles, 1, N_words, 'single');
        while length(current_ranking) > 1
            p_rr = p_rr .* ...
                p_x_particles_first(:, current_ranking(1), :) ./ ...
                sum(p_x_particles_first(:, current_ranking, :), 2);
            current_ranking = current_ranking(2:end);
        end
        p_x_rankings(:, rr, :) = p_rr;
    end
    clearvars current_ranking p_rr p_x_particles_first

    %% P_xy
    cols_to_repeat = kron(1:N_rankings, ones(1, temp_options));
    p_xy = p_x_rankings(:, cols_to_repeat, :) .* p_y;
    clearvars p_x_rankings p_y
    % Normalize
    p_xy = p_xy ./ (repmat(sum(p_xy, 2), [1, size(p_xy, 2), 1]));
    
    %% Heuristic
    % Compute the Entropy Ratios
    p_xy(p_xy<1e-10) = 1e-10;
    xylogxy_all = squeeze(sum(p_xy .* log2(p_xy), 2));
    E_H_particle = mean(-xylogxy_all);
    E_p_xy = squeeze(mean(p_xy)); 
    clearvars p_xy xylogxy_all
    E_p_xy(E_p_xy<1e-10) = 1e-10;
    H_E_particle = -sum(E_p_xy .* log2(E_p_xy), 1) ;
    H_ratio = H_E_particle - E_H_particle; % For N_prev 1 and 2 it is around 2.3, 2.0968, but for 3 it is 4.1 and 3.8
    [max_score, new_index] = max(H_ratio);
    word = list_embeddings(:, new_index);    
    list_embeddings(:, new_index) = [];
    new_index = new_index + sum(previous_index <= new_index);
    index = [previous_index, new_index];
    words = [previous_words, word];
end